firebase-perf/src/main/java/com/google/firebase/perf/config/DeviceCacheManager.java

```diff
@@ -242,21 +244,33 @@ public Optional<Float> getFloat(String key) {
    *
    * @return whether provided value has been saved to device caching layer.
    */
-  public boolean setValue(String key, float value) {
+  public boolean setValue(String key, double value) {
     if (key == null) {
-      logger.debug("Key is null when setting float value on device cache.");
+      logger.debug("Key is null when setting double value on device cache.");
       return false;
     }
+    key = newKeyForDoubleMigration(key);
+
     if (sharedPref == null) {
       setContext(getFirebaseApplicationContext());
       if (sharedPref == null) {
         return false;
       }
     }
-    sharedPref.edit().putFloat(key, value).apply();
+    sharedPref.edit().putLong(key, Double.doubleToRawLongBits(value)).apply();
     return true;
   }
 
+  /**
+   * This method creates a new key to disambiguate from an old one.
+   *
+   * <p>It should be used when we are changing the type of a key, so that we don't try to read the
+   * key with the old type and cause a crash.
+   */
+  private static String newKeyForDoubleMigration(String key) {
```

Comment:
Do we really need a new key? If the old value is recorded as float and the new value type we are looking up is Double:

1. Can we typecast to Double since Double should be able to handle the float precision?
2. To mitigate the potential exception thrown, can we bring in an exception handler to handle this in place?

If none of the above cases are true, then a new flag is fine. But I would want to avoid a flag transition phase if possible. Also, the new flag name does not look that friendly as a name :) (Huge thanks for abstracting :))

Reply:
I have changed the code to not use the new key. I think it's a bit more complicated now, but overall it should be fine.